consensus/istanbul/celo-ibft-spec.md

```diff
@@ -0,0 +1,394 @@
+  }
+  return 2f+1 <= |M| <= 3f+1 &&
+}
+```
+
+#### quorumRound
+
+Asserts that at least 2f+1 round change messages share the same round and
+returns that round.
+```
+quorumRound() {
+  M ← { m<RC_T, Hc, Rm, *>, n<RC_T, Hc, Rn, *> : Rm = Rn } &&
+  |M| >= 2f+1 &&
+  ∃ R : ∀ m<*, *, Rm, *> ∈ M : R = Rm
+  return R
+}
+```
+
+#### f1Round
+
+Asserts that there are at least f+1 round change messages and returns the
+lowest round from the top f+1 rounds.
+```
+f1Round() {
+  // This is saying that for any Rm there cannot be >= f+1 elements set with a
+  // larger R, since if there were that would mean that Rm is not in the top
+  // f+1 rounds. 
+  M ← { m<RC_T, Hc, Rm, *> : |{ n<RC_T, Hc, Rn, *> : Rm < Rn }| < f+1 } &&
+  |M| >= f+1 &&
+  ∃ R : ∀ m<*, *, Rm, *> ∈ M : R <= Rm
+  return R
+}
+```
+
+#### onRoundChangeTimeout
+As long as the round and height have not changed since it was scheduled
+onRoundChangeTimeout sets the desired round to be one greater than the Current
+round, sets the current state to be WaitingForNewRound and broadcasts a round
+change message.
+
+Note: This function is referred to in the code as
+`handleTimeoutAndMoveToNextRound`, which is misleading because it does not move
+to the next round, it only updates the desired round and sends a round change
+message. Hence why it has been renamed here to avoid confusion.
+
+```
+onRoundChangeTimeout(H, R) {
+  if H = Hc && R = Rc {
+    Rd ← Rc+1
+    Sc ← WaitingForNewRound
+    schedule onRoundChangeTimeout(Hc, Rd) after roundChangeTimeout(Rd)
+    bc<RC_T, Hc, Rd, PCc>
+  }
+}
+```
+
+
+## Appendix 1: Notation
+Elements of sets are represented with lower case letters (e.g. `m ∈ M`).
+Because all sets are messages we use `m` to represent an element and if we
+need to denote 2 messages from the same set we use `n` to denote the second
+message. Other variables are represented with Upper case letters (e.g. `H` for
+height).
+
+Composite objects are defined by a set of comma separated variables enclosed in
+angled brackets.\
+E.g. `<A, B, C>`
+
+If we need to refer to the composite object we prefix the angled brackets with
+an identifier that is a lower case letter.\
+E.g. `m<A, B, C>`
+
+An identifier can also be provided in order to distinguish variables belonging
+to a composite object from other similarly named variables in the same scope.
+In this case the variables are annotated with the identifier by appending the
+identifier to the variable name.\
+E.g. `m<Am, Bm, Cm>`
+
+If all instances of a composite object element share the same value for one
+of its variables then that value can be used in the definition. E.g. `<A, B, Hc>`
+represents a composite element with variables `A` `B` and current height.
+
+If a composite object element has variables for which the value is not
+important then `*` is used in the place of that variable.
+
+### Variable names
```

Comment:
While correct, I find myself going back and forth too often to this section to understand what these variables are. I feel like we can probably make them a little bit more self explanatory in many cases. Not just variable names. E.g "bc" could probably be completely renamed to 'broadcast' and an explanation line for it wouldn't be needed.

Reply:
There are 3 versions of this document, this one you have commented on with short names and then 2 others with long names. Can you check the others to see what you think?

Reply:
I'm wondering if it was obvious that variables mostly start with the first letter of the full name (e.g. Height -> H, Round -> R ...). And then there is the subscript notation for global variables either 'c' for current or 'd' desired.
With those two rules in my head I find it easy to remember all the notation, but of course I wrote it :D

I'm wondering if it would be better to put the notation up front and explain this naming approach. I can see how if you just dive into the algorithm and look up variables as you need, the naming approach may not be so clear.